File: ios/App/Settings/Devices/DevicesScreen.swift
// ---------------------------------------------------------------------------
// DevicesScreen — User-group "Devices" category page. One Signal card: linked →
// masked account + since-date + Unlink (confirm → refetch); unlinked → Link
// Signal → QR sheet (rendered from the gateway's qrDataUrl PNG) + status poll →
// linked/failed. Screen exit or Cancel stops the poll and fires link/cancel.
//
// Owns the @Observable DevicesViewModel via @State; DevicesBody is stateless
// (previewable per SignalState). The QR sheet's presentation is the VM's
// `isLinking`; dismissing it cancels the pairing.
// ---------------------------------------------------------------------------
import SwiftUI
import UIKit
import MobileData

struct DevicesScreen: View {
    @State private var vm: DevicesViewModel
    private let onBack: () -> Void

    init(settings: SettingsComponent, onBack: @escaping () -> Void) {
        self.onBack = onBack
        _vm = State(initialValue: DevicesViewModel(devices: settings.devices))
    }

    var body: some View {
        DevicesBody(
            signal: vm.signal,
            isUnlinking: vm.isUnlinking,
            onLink: { Task { await vm.startLink() } },
            onUnlink: { Task { await vm.unlink() } }
        )
        .task { await vm.load() }
        .onDisappear { vm.cancelLink() }
        .sheet(isPresented: linkingBinding) {
            QrLinkSheet(
                phase: vm.linkPhase,
                onRetry: { Task { await vm.startLink() } },
                onCancel: { vm.cancelLink() }
            )
        }
    }

    private var linkingBinding: Binding<Bool> {
        Binding(get: { vm.isLinking }, set: { open in if !open { vm.cancelLink() } })
    }
}

/// Stateless Signal card body; owns only the transient unlink-confirm toggle.
private struct DevicesBody: View {
    let signal: DevicesViewModel.SignalState
    let isUnlinking: Bool
    let onLink: () -> Void
    let onUnlink: () -> Void

    @State private var confirmingUnlink = false

    var body: some View {
        SettingsPageScaffold(title: "Devices", screenId: "settings-devices-screen") {
            SettingsCard(title: "Signal", sub: "Text-chat via your own Signal account.") {
                content
            }
        }
        .confirmationDialog("Disconnect Signal?", isPresented: $confirmingUnlink, titleVisibility: .visible) {
            Button("Disconnect", role: .destructive, action: onUnlink)
            Button("Cancel", role: .cancel) {}
        } message: {
            Text("This removes Sentient as a linked device from your Signal account. Past conversation history stays in your memory.")
        }
    }

    @ViewBuilder
    private var content: some View {
        switch signal {
        case .loading:
            ProgressView().controlSize(.small).padding(.vertical, Space.md)
        case .linked(let account, let since):
            linkedView(account: account, since: since)
        case .unlinked:
            unpairedView
        case .loadError:
            Text("Couldn't load device status.")
                .font(Typo.ui(TypeScale.sm))
                .foregroundStyle(DuskColors.ink3)
                .padding(.vertical, Space.md)
        }
    }

    private func linkedView(account: String?, since: String?) -> some View {
        VStack(alignment: .leading, spacing: Space.md) {
            HStack(spacing: Space.sm) {
                Circle().fill(DuskColors.ok).frame(width: 8, height: 8)
                Text("Linked").font(Typo.ui(TypeScale.sm, .semibold)).foregroundStyle(DuskColors.ink)
                if let account { Text(account).font(Typo.mono(TypeScale.sm)).foregroundStyle(DuskColors.ink2) }
                if let since { Text("since \(formatSince(since))").font(Typo.ui(TypeScale.xs)).foregroundStyle(DuskColors.ink3) }
            }
            DangerButton(title: isUnlinking ? "Unlinking…" : "Unlink", accessibilityId: "settings-devices-unlink") {
                confirmingUnlink = true
            }
            .disabled(isUnlinking)
        }
        .padding(.vertical, Space.sm)
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private var unpairedView: some View {
        VStack(alignment: .leading, spacing: Space.md) {
            Text("Text-chat with your agent from Signal. Links your own Signal account using \"Note to Self\" — no second number needed.")
                .font(Typo.ui(TypeScale.sm))
                .foregroundStyle(DuskColors.ink2)
            Button("Link Signal", action: onLink)
                .font(Typo.ui(TypeScale.sm, .semibold))
                .foregroundStyle(DuskColors.accent)
                .accessibilityIdentifier("settings-devices-link")
        }
        .padding(.vertical, Space.sm)
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

/// The link sheet: scan-steps + QR (or preparing / error) + Cancel.
private struct QrLinkSheet: View {
    let phase: DevicesViewModel.LinkPhase
    let onRetry: () -> Void
    let onCancel: () -> Void

    private static let qrSide: CGFloat = 240

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: Space.lg) {
                    steps
                    qrArea
                }
                .padding(.horizontal, Space.lg)
                .padding(.top, Space.lg)
                .frame(maxWidth: .infinity, alignment: .leading)
            }
            .background(DuskColors.bg)
            .navigationTitle("Link your Signal account")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel", action: onCancel)
                        .accessibilityIdentifier("settings-devices-link-cancel")
                }
            }
            .duskTheme()
        }
    }

    private var steps: some View {
        VStack(alignment: .leading, spacing: Space.xs) {
            stepLine("1. On your phone, open Signal")
            stepLine("2. Tap your avatar → Linked Devices")
            stepLine("3. Tap \"Link New Device\"")
            stepLine("4. Scan this QR code")
        }
    }

    private func stepLine(_ text: String) -> some View {
        Text(text).font(Typo.ui(TypeScale.sm)).foregroundStyle(DuskColors.ink2)
    }

    @ViewBuilder
    private var qrArea: some View {
        switch phase {
        case .idle, .preparing:
            HStack { Spacer(); ProgressView("Preparing link…"); Spacer() }
                .padding(.vertical, Space.xl)
        case .active(let qr):
            qrImage(qr)
        case .failed(let message):
            VStack(spacing: Space.md) {
                Text(message).font(Typo.ui(TypeScale.sm, .medium)).foregroundStyle(DuskColors.stop)
                Button("Retry", action: onRetry)
                    .font(Typo.ui(TypeScale.sm, .semibold))
                    .foregroundStyle(DuskColors.accent)
                    .accessibilityIdentifier("settings-devices-link-retry")
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, Space.lg)
        }
    }

    @ViewBuilder
    private func qrImage(_ qr: Data) -> some View {
        if let image = UIImage(data: qr) {
            HStack {
                Spacer()
                Image(uiImage: image)
                    .interpolation(.none)
                    .resizable()
                    .frame(width: Self.qrSide, height: Self.qrSide)
                    .background(Color.white)
                    .clipShape(RoundedRectangle(cornerRadius: Radii.sm))
                    .accessibilityIdentifier("settings-devices-qr")
                Spacer()
            }
        } else {
            Text("Couldn't render the linking code.")
                .font(Typo.ui(TypeScale.sm))
                .foregroundStyle(DuskColors.stop)
        }
    }
}

/// ISO-8601 → short local date; falls back to the raw string on parse failure.
private func formatSince(_ iso: String) -> String {
    guard let date = ISO8601DateFormatter().date(from: iso) else { return iso }
    let formatter = DateFormatter()
    formatter.dateStyle = .medium
    formatter.timeStyle = .none
    return formatter.string(from: date)
}

// ── Previews — Signal states (no VM) ─────────────────────────────────────────

#Preview("linked") {
    NavigationStack {
        DevicesBody(
            signal: .linked(account: "+1 •••• ••34", since: "2026-06-01T12:00:00Z"),
            isUnlinking: false, onLink: {}, onUnlink: {}
        )
    }
    .preferredColorScheme(.dark)
}

#Preview("unlinked") {
    NavigationStack {
        DevicesBody(signal: .unlinked, isUnlinking: false, onLink: {}, onUnlink: {})
    }
    .preferredColorScheme(.dark)
}

File: android/src/main/kotlin/io/sentient/android/settings/devices/DevicesScreen.kt
// ---------------------------------------------------------------------------
// DevicesScreen — the Devices settings page (User group). One Signal card that is
// either Linked (masked account + linked date + Unlink w/ confirm) or Unlinked
// ("Link Signal" → QR + status poll). All state + callbacks come from
// [DevicesViewModel]; leaf views are pure. testTags: settings-devices-screen/-back,
// settings-devices-{link,unlink,unlink-confirm,unlink-cancel}.
// ---------------------------------------------------------------------------
package io.sentient.android.settings.devices

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeDrawingPadding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import io.sentient.android.settings.components.DangerButton
import io.sentient.android.settings.components.SettingsCard
import io.sentient.android.settings.components.SettingsTopBar
import io.sentient.android.theme.LocalTokens
import io.sentient.android.theme.SentientTheme
import io.sentient.mobilesdk.design.Colors

private const val TITLE = "Devices"
private val DOT_SIZE = 8.dp

@Composable
fun DevicesScreen(
    vm: DevicesViewModel,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val state by vm.state.collectAsStateWithLifecycle()
    DevicesContent(
        state = state,
        onBack = onBack,
        onLink = vm::startLink,
        onCancelLink = vm::cancelLink,
        onRetryLink = vm::retryLink,
        onUnlinkClick = vm::openUnlinkConfirm,
        onUnlinkConfirm = vm::unlink,
        onUnlinkCancel = vm::closeUnlinkConfirm,
        modifier = modifier,
    )
}

@Composable
private fun DevicesContent(
    state: DevicesUiState,
    onBack: () -> Unit,
    onLink: () -> Unit,
    onCancelLink: () -> Unit,
    onRetryLink: () -> Unit,
    onUnlinkClick: () -> Unit,
    onUnlinkConfirm: () -> Unit,
    onUnlinkCancel: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val tokens = LocalTokens.current
    Column(
        modifier = modifier.fillMaxSize().safeDrawingPadding().testTag("settings-devices-screen"),
    ) {
        SettingsTopBar(title = TITLE, onBack = onBack, backTestTag = "settings-devices-back")
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = tokens.space.lg, vertical = tokens.space.md),
            verticalArrangement = Arrangement.spacedBy(tokens.space.md),
        ) {
            SettingsCard(title = "Signal", subtitle = "Text-chat via your own Signal account.") {
                Column(
                    modifier = Modifier.padding(horizontal = tokens.space.lg, vertical = tokens.space.sm),
                    verticalArrangement = Arrangement.spacedBy(tokens.space.sm),
                ) {
                    when (val card = state.card) {
                        SignalCardState.Loading ->
                            Text("Loading…", color = Color(Colors.ink3), fontSize = tokens.type.sm)
                        is SignalCardState.Linked -> LinkedView(card, state.busy, onUnlinkClick)
                        SignalCardState.Unlinked -> UnlinkedView(onLink)
                    }
                    if (state.errorMessage != null) {
                        Text(
                            state.errorMessage,
                            modifier = Modifier.testTag("settings-devices-error"),
                            color = Color(Colors.stop),
                            fontSize = tokens.type.sm,
                        )
                    }
                }
            }
        }
    }

    SignalLinkDialog(link = state.link, onCancel = onCancelLink, onRetry = onRetryLink)

    if (state.unlinkConfirmOpen) {
        AlertDialog(
            onDismissRequest = onUnlinkCancel,
            title = { Text("Disconnect Signal?") },
            text = {
                Text(
                    "This removes Sentient as a linked device from your Signal account. " +
                        "Past Signal conversation history is kept in your Sentient memory.",
                    color = Color(Colors.ink3),
                    fontSize = tokens.type.sm,
                )
            },
            confirmButton = {
                TextButton(
                    onClick = onUnlinkConfirm,
                    enabled = !state.busy,
                    modifier = Modifier.testTag("settings-devices-unlink-confirm"),
                ) { Text("Disconnect", color = Color(Colors.stop)) }
            },
            dismissButton = {
                TextButton(
                    onClick = onUnlinkCancel,
                    enabled = !state.busy,
                    modifier = Modifier.testTag("settings-devices-unlink-cancel"),
                ) { Text("Cancel") }
            },
        )
    }
}

@Composable
private fun LinkedView(card: SignalCardState.Linked, busy: Boolean, onUnlink: () -> Unit) {
    val tokens = LocalTokens.current
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(tokens.space.sm),
    ) {
        Box(modifier = Modifier.size(DOT_SIZE).clip(CircleShape).background(Color(Colors.ok)))
        Text("Linked", color = Color(Colors.ink), fontSize = tokens.type.base)
        card.accountMasked?.let { Text(it, color = Color(Colors.ink2), fontSize = tokens.type.sm) }
    }
    card.linkedAt?.let {
        Text("Linked ${formatLinkedDate(it)}", color = Color(Colors.ink3), fontSize = tokens.type.xs)
    }
    Text(
        "Open your \"Note to Self\" thread in Signal and text your agent like any conversation.",
        color = Color(Colors.ink3),
        fontSize = tokens.type.sm,
    )
    DangerButton(
        label = "Unlink",
        onClick = onUnlink,
        enabled = !busy,
        testTag = "settings-devices-unlink",
    )
}

@Composable
private fun UnlinkedView(onLink: () -> Unit) {
    val tokens = LocalTokens.current
    Text(
        "Text-chat with your agent from Signal. Uses \"Note to Self\" mode — links your own " +
            "Signal account, no second number needed.",
        color = Color(Colors.ink3),
        fontSize = tokens.type.sm,
    )
    Button(onClick = onLink, modifier = Modifier.testTag("settings-devices-link")) {
        Text("Link Signal")
    }
}

/** Show the date portion of an ISO-8601 linked_at (YYYY-MM-DD), or the raw string if shorter. */
private fun formatLinkedDate(linkedAt: String): String =
    if (linkedAt.length >= 10) "on ${linkedAt.take(10)}" else linkedAt

@Preview
@Composable
private fun DevicesLinkedPreview() {
    SentientTheme {
        DevicesContent(
            state = DevicesUiState(card = SignalCardState.Linked("+1 •••• 1234", "2026-07-16T10:00:00Z")),
            onBack = {}, onLink = {}, onCancelLink = {}, onRetryLink = {},
            onUnlinkClick = {}, onUnlinkConfirm = {}, onUnlinkCancel = {},
        )
    }
}

@Preview
@Composable
private fun DevicesUnlinkedPreview() {
    SentientTheme {
        DevicesContent(
            state = DevicesUiState(card = SignalCardState.Unlinked),
            onBack = {}, onLink = {}, onCancelLink = {}, onRetryLink = {},
            onUnlinkClick = {}, onUnlinkConfirm = {}, onUnlinkCancel = {},
        )
    }
}
